### PR TITLE
DOCSP-27376: removed support for dot notation in find*

### PR DESCRIPTION
**Pull request by rustagir**
# Pull Request Info

[PR Reviewing Guidelines](https://github.com/mongodb/docs-node/blob/master/REVIEWING.md)

JIRA - https://jira.mongodb.org/browse/DOCSP-27376
Staging - https://docs-mongodbcom-staging.corp.mongodb.com/node/docsworker-xlarge/DOCSP-27376-dot-notation-unsupported/fundamentals/typescript/#type-safety-and-dot-notation
https://docs-mongodbcom-staging.corp.mongodb.com/node/docsworker-xlarge/DOCSP-27376-dot-notation-unsupported/whats-new/#what-s-new-in-5.0

## Self-Review Checklist

- [ ] Is this free of any warnings or errors in the RST?
- [ ] Did you run a spell-check?
- [ ] Did you run a grammar-check?
- [ ] Are all the links working?

### CODE REVIEW

File: source/fundamentals/typescript.txt [hunk 1]
```diff
@@ -104,22 +106,53 @@ genus and color of dogs and cats:
...
 
-   No overload matches this call.
-   ...
-   Type '"Sylvilagus"' is not assignable to type 'Condition<"Canis" | "Felis">'.
+   Type 'number' is not assignable to type 'string'.
+   
+Dot Notation in Find Operations
+```````````````````````````````
+
+Starting in version 5.0, the {+driver-short+} does not provide type
```
Comment (mongoKart) [inline source/fundamentals/typescript.txt:126]:
```suggestion
Starting in version 5.0, by default, the {+driver-short+} does not provide type
```

File: source/fundamentals/typescript.txt [hunk 2]
```diff
@@ -104,22 +106,53 @@ genus and color of dogs and cats:
...
+
+The following code compiles, even though the value of
+``classification.color`` should be a string:
+
+.. code-block:: typescript
+
+   await collection.find({ "classification.color": false })
+
+Even though support for this checking is removed by default, you can
+enable type-checking by constructing filters as ``StrictFilter`` types.
```
Comment (mongoKart) [inline source/fundamentals/typescript.txt:141]: I: We probably want to mention that this is the default setting earlier in the section (see comment above). With that done, we can simplify this sentence.

```suggestion
You can enable type safety for fields in dot notation by constructing filters as ``StrictFilter`` types.
```

File: source/fundamentals/typescript.txt [hunk 3]
```diff
@@ -104,22 +106,53 @@ genus and color of dogs and cats:
...
+
+.. warning::
+   
+   Use of the ``StrictFilter`` type to create filters is experimental and currently has
+   unknown performance implications.
+
+The following code does not compile because the filter is assigned as a
+``StrictFilter`` type, which causes the driver to report a type error
+because the value of ``classification.color`` should be a string:
```
Comment (mongoKart) [inline source/fundamentals/typescript.txt:150]: I: Sentence is a little long

```suggestion
In the following code sample, the search filter is assigned a
``StrictFilter`` type. Because the value of ``classification.color`` should be a string,
the {+driver-short+} reports a type error when a boolean is used:
```

File: source/whats-new.txt [hunk 4]
```diff
@@ -41,6 +41,13 @@ What's New in 5.0
 
 New features of the 5.0 {+driver-short+} release include:
 
+- Removal of default Typescript type-checking of types expressed in dot
+  notation.
+
+  - The driver no longer checks types referenced in dot notation during
+    find operations unless the ``StrictFilter`` feature is explicitly
+    implemented. To learn more about this change, see the :ref:`Typescript fundamentals page <node-ts-type-safety>`.
```
Comment (mongoKart) [inline source/whats-new.txt:49]:
```suggestion
  - By default, the driver no longer checks types referenced in dot notation during
    find operations unless the ``StrictFilter`` feature is explicitly
    implemented. To learn more about this change, see the :ref:`Typescript fundamentals page <node-ts-type-safety>`.
```

**Review by mongoKart — CHANGES_REQUESTED** [review]
A few suggestions to help clarify a hard-to-articulate idea. Let me know your thoughts!

File: source/fundamentals/typescript.txt [hunk 5]
```diff
@@ -104,22 +106,53 @@ genus and color of dogs and cats:
...
+``find()`` or ``findOneAndUpdate()``, your code compiles even if you
+specify a value of an incorrect type for a field expressed in dot
+notation.
+
+The following code compiles, even though the value of
+``classification.color`` should be a string:
+
+.. code-block:: typescript
+
+   await collection.find({ "classification.color": false })
```
[the remark this answers is not in the document]
Comment (rustagir) [reply source/fundamentals/typescript.txt:138]: Sure! Updated to `findOneAndDelete()`

File: source/fundamentals/typescript.txt [hunk 6]
```diff
@@ -104,22 +106,53 @@ genus and color of dogs and cats:
...
+   
+Dot Notation in Find Operations
+```````````````````````````````
+
+Starting in version 5.0, the {+driver-short+} does not provide type
+safety for *find operations* that search on fields expressed in dot
+notation. When you construct a filter to pass to a find method, such as
+``find()`` or ``findOneAndUpdate()``, your code compiles even if you
+specify a value of an incorrect type for a field expressed in dot
+notation.
```
[the remark this answers is not in the document]
Comment (rustagir) [reply source/fundamentals/typescript.txt:131]: moved list of methods to admonition and reworded for clarity.

File: source/fundamentals/typescript.txt [hunk 7]
```diff
@@ -104,22 +106,53 @@ genus and color of dogs and cats:
...
+
+The following code compiles, even though the value of
+``classification.color`` should be a string:
```
[the remark this answers is not in the document]
Comment (rustagir) [reply source/fundamentals/typescript.txt:134]: how do you feel about the way I reworded this instead?

File: source/fundamentals/typescript.txt (hunk 3 above)
Comment (rustagir) [reply source/fundamentals/typescript.txt:150]: Reworded.

File: source/fundamentals/typescript.txt [hunk 8]
```diff
@@ -123,36 +123,46 @@ The type error raised by the preceding code snippet is as follows:
...
+safety for find operations that search on fields expressed in dot
+notation. When you construct a filter to pass to a find method, the
+driver will not raise a type error even if you specify an incorrectly
+typed value for a field expressed in dot notation.
 
-The following code compiles, even though the value of
-``classification.color`` should be a string:
+.. note::
+
+   The following methods comprise find operations:
```
Comment (mongoKart) [inline source/fundamentals/typescript.txt:134]: I: [Style Guide recommends not using "comprise" at all](https://www.mongodb.com/docs/meta/style-guide/terminology/alphabetical-terms/#c) (even using it correctly, which is weird)

S:
```suggestion
   The following methods are considered find operations:
```
Or:
```suggestion
   The following methods are find operations:
```
Or something similar.

**Review by mongoKart — APPROVED** [review]
Nice job. One small line to reword.

File: source/fundamentals/typescript.txt [hunk 9]
```diff
@@ -104,22 +106,63 @@ genus and color of dogs and cats:
...
+
+.. note::
+
+   The following methods comprise find operations:
+
+   - ``find()``
+   - ``findOne()``
+   - ``findOneAndDelete()``
+   - ``findOneAndReplace()``
+   - ``findOneAndUpdate()``
```
Comment (ccho-mongodb) [inline source/fundamentals/typescript.txt:140]: Suggestion:

For consistency, we should standardize "find operations" to match this list for all drivers. For example, the golang docs only mention find and findOne:
https://github.com/mongodb/docs-golang/blob/master/source/fundamentals/crud/read-operations/retrieve.txt#L59-L60

Comment (rustagir) [reply source/fundamentals/typescript.txt:140]: changed the language for this list, but agreed- this should be standardized. Making a ticket for myself.

File: source/whats-new.txt [hunk 10]
```diff
@@ -41,6 +41,10 @@ What's New in 5.0
 
 New features of the 5.0 {+driver-short+} release include:
 
+- By default, the driver no longer checks types referenced in dot notation
+  unless the ``StrictFilter`` feature is explicitly
+  implemented. To learn more about this change, see the :ref:`Typescript fundamentals page <node-ts-type-safety>`.
+
 - Optional installation of ``@aws-sdk/credential-providers`` as a dependency.
 
   - The driver no longer includes AWS SDK modules by default. Use the
```
Comment (nbbeeken) [inline source/whats-new.txt:50]: Sorry this is unrelated to the work here, the aws sdk version below isn't correct, and some shells don't work well with `@` and caret symbols `^` (which we should also add to the version) so quoting it is a good idea for maximum compat:

```bash
npm install --save "@aws-sdk/credential-providers@^3.201.0"
```

File: source/whats-new.txt [hunk 11]
```diff
@@ -41,6 +41,10 @@ What's New in 5.0
 
 New features of the 5.0 {+driver-short+} release include:
 
+- By default, the driver no longer checks types referenced in dot notation
+  unless the ``StrictFilter`` feature is explicitly
+  implemented. To learn more about this change, see the :ref:`Typescript fundamentals page <node-ts-type-safety>`.
```
Comment (nbbeeken) [inline source/whats-new.txt:46]:
```suggestion
- By default, the driver no longer checks types referenced in dot notation
  unless the ``StrictFilter`` type annotation is explicitly
  used. To learn more about this change, see the :ref:`Typescript fundamentals page <node-ts-type-safety>`.
```

File: source/fundamentals/typescript.txt [hunk 12]
```diff
@@ -104,22 +108,30 @@ genus and color of dogs and cats:
...
-     .find({ "classification.genus": "Sylvilagus" });
+   await collection.findOneAndDelete({ "classification.color": false })
+
+You can enable type-checking by constructing filters as ``StrictFilter`` types.
 
-The type error raised by the preceding code snippet is as follows:
+.. warning::
+   
+   Use of the ``StrictFilter`` type to create filters is experimental and currently has
+   unknown performance implications.
```
Comment (nbbeeken) [inline source/fundamentals/typescript.txt:124]:
```suggestion
   Use of the ``StrictFilter`` type to create filters is experimental and may have type
   errors in places where there should be none.
```

We aren't worried about the performance, it is more about the usability hurdles that can get in the way of writing a "correct query" from the server's perspective but that our typescript type can't handle.

File: source/fundamentals/typescript.txt [hunk 13]
```diff
@@ -104,22 +108,30 @@ genus and color of dogs and cats:
...
+reports a type error because the value of ``classification.color`` is a
+boolean instead of a string:
+
+.. code-block:: typescript
 
-   No overload matches this call.
-   ...
-   Type '"Sylvilagus"' is not assignable to type 'Condition<"Canis" | "Felis">'.
+   const filterPredicate: StrictFilter<ClassificationPet> = { "classification.color": false };
+   await collection.findOneAndDelete(filterPredicate);
```
Comment (nbbeeken) [inline source/fundamentals/typescript.txt:134]: We also had dot-notation support for UpdateFilter now moved to StrictUpdateFilter, I'm not sure if we need code snippets. It would be the same kind of migration for those that wanted to get back the strict behavior (which is also experimental):

```ts
// v4.x, compiler error
pets.updateOne({}, { $set: { "classification.color": false } })
// v5.x, no error
pets.updateOne({}, { $set: { "classification.color": false } })
// v5.x, compiler error
const update: StrictUpdateFilter<ClassificationPet> = { $set: { "classification.color": false } }
pets.updateOne({}, update)
```

File: source/whats-new.txt (hunk 11 above)
Comment (nbbeeken) [reply source/whats-new.txt:46]: Additionally, in our migration guide we made a **bold header call out** that this is a TypeScript only difference that does not effect runtime database queries/operations. I'm sure you have your own style rules but we felt it's important that there isn't confusion that our driver does special handling of keys with dots in them at runtime.

**Review by nbbeeken — CHANGES_REQUESTED** [review]

**Review by nbbeeken — APPROVED** [review]
LGTM
